nodejs/helpers.js

```diff
@@ -0,0 +1,172 @@
+const fs = require("fs")
+const mongodb = require("mongodb")
+const { ClientEncryption } = require("mongodb-client-encryption")(mongodb)
+const { MongoClient, Binary } = mongodb
+
+module.exports = {
+  readMasterKey: function(path = "./master-key.txt") {
+    return fs.readFileSync(path)
+  },
+  CsfleHelper: class {
+    constructor({
+      kmsProviders = null,
+      keyAltNames = "demo-data-key",
+      keyDB = "encryption",
+      keyColl = "__keyVault",
+      schema = null,
+      connectionString = "mongodb://localhost:27017",
+      mongocryptdBypassSpawn = false,
+      mongocryptdSpawnPath = "mongocryptd"
+    } = {}) {
+      if (kmsProviders === null) {
+        throw new Error("kmsProviders is required")
+      }
+      this.kmsProviders = kmsProviders
+      this.keyAltNames = keyAltNames
+      this.keyDB = keyDB
+      this.keyColl = keyColl
+      this.keyVaultNamespace = `${keyDB}.${keyColl}`
+      this.schema = schema
+      this.connectionString = connectionString
+      this.mongocryptdBypassSpawn = mongocryptdBypassSpawn
+      this.mongocryptdSpawnPath = mongocryptdSpawnPath
+      this.regularClient = null
+      this.csfleClient = null
+    }
+
+    /**
+     * Creates a unique, partial index in the key vault collection
+     * on the ``keyAltNames`` field.
+     *
+     * @param {MongoClient} client
+     */
+    async ensureUniqueIndexOnKeyVault(client) {
+      try {
+        await client
+          .db(this.keyDB)
+          .collection(this.keyColl)
+          .createIndex("keyAltNames", {
+            unique: true,
+            partialFilterExpression: {
+              keyAltNames: {
+                $exists: true
+              }
+            }
+          })
+      } catch (e) {
+        console.error(e)
+        process.exit(1)
+      }
+    }
+
+    /**
+     * In the guide, https://docs.mongodb.com/ecosystem/use-cases/client-side-field-level-encryption-guide/,
+     * we create the data key and then show that it is created by
+     * retreiving it using a findOne query. Here, in implementation, we only
+     * create the key if it doesn't already exist, ensuring we only have one
+     * local data key.
+     */
+    async findOrCreateDataKey() {
```

Comment:
What I was saying before is just that it seems like you don't need to make another client here. The client used with `ClientEncryption` is just a regular, non-autoencrypted client, and I think you already make one of those in all cases. I would probably rewrite this method to take in a `client: MongoClient`, rather than creating a one-off client here. Along the lines of earlier critiques, I think it's hiding what's actually going on (another client is being created), and will ultimately likely confuse the users. 

At very least it looks like you could replace this code with:
```
const client = await getRegularClient();
```

Reply:
ah, I see what you mean.